mixer/pkg/runtime2/handler/table_test.go

```diff
@@ -164,6 +172,81 @@ func TestCleanup_Basic(t *testing.T) {
 	}
 }
 
+func TestCleanup_WorkerNotClosed(t *testing.T) {
+	tests := []struct {
+		SpawnWorker            bool
+		SpawnDaemon            bool
+		CloseGoRoutines        bool
+		wantWorkerStrayRoutine float64
+		wantDaemonStrayRoutine float64
+	}{
+
+		{
+			SpawnWorker:            true,
+			SpawnDaemon:            true,
+			CloseGoRoutines:        false,
+			wantWorkerStrayRoutine: 1,
+			wantDaemonStrayRoutine: 1,
+		},
+		{
+			SpawnWorker:            true,
+			SpawnDaemon:            true,
+			CloseGoRoutines:        true,
+			wantWorkerStrayRoutine: 0,
+			wantDaemonStrayRoutine: 0,
+		},
+		{
+			SpawnWorker:            true,
+			SpawnDaemon:            false,
+			CloseGoRoutines:        false,
+			wantWorkerStrayRoutine: 1,
+			wantDaemonStrayRoutine: 0,
+		},
+		{
+			SpawnWorker:            false,
+			SpawnDaemon:            true,
+			CloseGoRoutines:        false,
+			wantWorkerStrayRoutine: 0,
+			wantDaemonStrayRoutine: 1,
+		},
+	}
+	for idx, tt := range tests {
+		t.Run(fmt.Sprintf("[%d]", idx), func(t *testing.T) {
+			adapters := data.BuildAdapters(data.FakeAdapterSettings{Name: "acheck", SpawnWorker: tt.SpawnWorker, SpawnDaemon: tt.SpawnDaemon, CloseGoRoutines: tt.CloseGoRoutines})
+			templates := data.BuildTemplates()
+
+			s := util.GetSnapshot(templates, adapters, data.ServiceConfig, globalCfg)
+			s.ID = int64(idx * 2)
+
+			oldTable := NewTable(Empty(), s, pool.NewGoroutinePool(5, false))
+
+			s = config.Empty()
+			s.ID = int64(idx*2 + 1)
+
+			newTable := NewTable(oldTable, s, nil)
+
+			oldTable.Cleanup(newTable)
+
+			// give time for counters to get updated before validating them.
+			time.Sleep(500 * time.Millisecond)
+
+			var c prometheus.Metric = oldTable.entries["hcheck1.acheck.istio-system"].env.counters.workers
```

Comment:
Nit: there should be a data.FqnHCheck1 type for capturing the fully qualified name of that entry.